networking/socket_network.go

```diff
@@ -0,0 +1,231 @@
+
+	"student_25_adkg/transport"
+)
+
+type TransportNetwork struct {
+	mu        sync.Mutex
+	transport transport.Transport
+	nextID    int64
+	nodes     map[int64]string // id -> address
+	sockets   map[int64]transport.Socket
+	peers     *PeerMap
+}
+
+// NewNetwork creates a new network instance using the given transport
+func NewTransportNetwork(t transport.Transport) *TransportNetwork {
+	return &TransportNetwork{
+		transport: t,
+		nextID:    1,
+		nodes:     make(map[int64]string),
+		sockets:   make(map[int64]transport.Socket),
+		peers:     NewPeerMap(),
+	}
+}
+
+type PeerMap struct {
+	sync.RWMutex
+	peers map[int64]string
+}
+
+func NewPeerMap() *PeerMap {
+	return &PeerMap{peers: make(map[int64]string)}
+}
+
+func (pm *PeerMap) GetPeers(excludeID int64) map[int64]string {
+	pm.RLock()
+	defer pm.RUnlock()
+
+	result := make(map[int64]string)
+	for id, addr := range pm.peers {
+		if id != excludeID {
+			result[id] = addr
+		}
+	}
+	return result
+}
+
+func (pm *PeerMap) Add(id int64, addr string) {
+	pm.Lock()
+	defer pm.Unlock()
+	pm.peers[id] = addr
+}
+
+func (n *TransportNetwork) JoinNetwork() (NetworkInterface, error) {
+	bindAddr := "127.0.0.1:0"
+	n.mu.Lock()
+	defer n.mu.Unlock()
+
+	socket, err := n.transport.CreateSocket(bindAddr)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create socket: %w", err)
+	}
+
+	id := n.nextID
+	n.nextID++
+	addr := socket.GetAddress()
+
+	n.sockets[id] = socket
+	n.peers.Add(id, addr)
+
+	return NewSocketNetwork(socket, id, n.peers), nil
+}
+
+type SocketNetwork struct {
+	socket transport.Socket
+	id     int64
+	// peers     map[int64]string // mapping from ID to address
+	peers     *PeerMap
+	incoming  chan []byte
+	stop      chan struct{}
+	recvWg    sync.WaitGroup
+	recvMutex sync.Mutex
+	sent      [][]byte
+	received  [][]byte
+}
+
+// NewSocketNetwork creates a new SocketNetwork with a given socket and ID.
+// `peers` maps IDs to addresses (host:port).
+func NewSocketNetwork(socket transport.Socket, id int64, peers *PeerMap) *SocketNetwork {
+
+	net := &SocketNetwork{
+		socket:   socket,
+		id:       id,
+		peers:    peers,
+		incoming: make(chan []byte, 100),
+		stop:     make(chan struct{}),
+		sent:     make([][]byte, 0),
+		received: make([][]byte, 0),
+	}
+
+	net.recvWg.Add(1)
+	go net.receiver()
+
+	return net
+}
+
+func (n *SocketNetwork) Send(msg []byte, to int64) error {
+	addrMap := n.peers.GetPeers(n.id)
+	addr, ok := addrMap[to]
+	if !ok {
+		return fmt.Errorf("unknown peer ID: %d", to)
+	}
+
+	header := transport.NewHeader(n.socket.GetAddress(), addr, addr)
+	packet := transport.Packet{
+		Header: &header,
+		Msg:    &transport.Message{Type: "bytes", Payload: msg},
+	}
+
+	timeout := time.Second
+	err := n.socket.Send(addr, packet, timeout)
+	if err != nil {
+		return err
+	}
+
+	n.recvMutex.Lock()
+	n.sent = append(n.sent, msg)
+	n.recvMutex.Unlock()
+
+	return nil
+}
+
+func (n *SocketNetwork) Broadcast(msg []byte) error {
+	addrMap := n.peers.GetPeers(n.id)
```

Comment:
Does not broadcast to itself because `GetPeers(n.id)` removes itself from the list. Use -1 instead.

```suggestion
	addrMap := n.peers.GetPeers(int64(-1))
```

Reply:
Yes, true, I've changed this and a couple of things like GetReceived() in my aba branch but did not commit here. I'll try to merge it with your changes without breaking anything, look into HACKERALERTs in go.sum and ping you when I commit it, it should be fine.

Good question about copying files into our repo. I'm not sure how it can be done as a dependency. Btw I've also deleted some internal transport tests which pulled other peerster dependencies and excluded the whole transport directory from linter checks supposing that it has been tested well enough.

My question would be how should we mention that transport has been borrowed from peerster? Is a comment `// COPIED FROM DEDIS CS438 course materials` in ./transport/transport.go enough?

Reply:
Yes no need to do more in the way of attribution. You have our permission :-)